File: python/ratelimiter/algorithms/token_bucket.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

from ..models import Decision, ScriptExecutionError
from .base import BasePolicy, RateLimitAlgorithm


@dataclass(frozen=True)
class TokenBucketPolicy(BasePolicy):
    capacity: int
    refill_rate: float
    ttl_sec: int = 60

    def validate(self) -> None:
        if self.capacity <= 0:
            raise ValueError("capacity must be > 0")
        if self.refill_rate < 0:
            raise ValueError("refill_rate must be >= 0")
        if self.ttl_sec <= 0:
            raise ValueError("ttl_sec must be > 0")


class TokenBucketAlgorithm(RateLimitAlgorithm):
    name = "token_bucket"
    key_prefix = "ratelimit:token_bucket"
    policy_type = TokenBucketPolicy

    def default_script_path(self) -> Path:
        return Path(__file__).resolve().parents[3] / "scripts" / "lua" / "token_bucket.lua"

    def validate_policy(self, policy: BasePolicy) -> None:
        if not isinstance(policy, TokenBucketPolicy):
            raise TypeError("token_bucket algorithm requires TokenBucketPolicy")
        policy.validate()

    def build_redis_args(self, policy: BasePolicy, requested: int) -> list[Any]:
        assert isinstance(policy, TokenBucketPolicy)
        return [policy.capacity, policy.refill_rate, requested, policy.ttl_sec]

    def parse_decision(self, raw: Any) -> Decision:
        if not isinstance(raw, (list, tuple)) or len(raw) != 4:
            raise ScriptExecutionError(f"unexpected script response: {raw!r}")

        allowed, remaining, retry_after_ms, reset_at_ms = raw
        return Decision(
            allowed=bool(int(allowed)),
            remaining=int(remaining),
            retry_after_ms=int(retry_after_ms),
            reset_at_ms=int(reset_at_ms),
        )

File: go/ratelimiter/algorithms/token_bucket.go
package algorithms

import (
	"fmt"
	"path/filepath"
	"runtime"
)

// TokenBucketPolicy configures the token-bucket algorithm.
type TokenBucketPolicy struct {
	Capacity   int64
	RefillRate float64
	TTLSec     int64
}

// Validate verifies token-bucket policy invariants.
func (p TokenBucketPolicy) Validate() error {
	if p.Capacity <= 0 {
		return fmt.Errorf("capacity must be > 0")
	}
	if p.RefillRate < 0 {
		return fmt.Errorf("refill_rate must be >= 0")
	}
	if p.TTLSec <= 0 {
		return fmt.Errorf("ttl_sec must be > 0")
	}
	return nil
}

// TokenBucketAlgorithm implements the Redis Lua token-bucket contract.
type TokenBucketAlgorithm struct{}

func (TokenBucketAlgorithm) Name() string { return "token_bucket" }

func (TokenBucketAlgorithm) KeyPrefix() string { return "ratelimit" }

func (TokenBucketAlgorithm) BuildRedisKeys(baseKey string) []string { return []string{baseKey} }

func (TokenBucketAlgorithm) UseRedisHashTag() bool { return false }

func (TokenBucketAlgorithm) DefaultScriptPath() string {
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		return filepath.Join("..", "..", "..", "scripts", "lua", "token_bucket.lua")
	}
	return filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", "..", "scripts", "lua", "token_bucket.lua"))
}

func (TokenBucketAlgorithm) ValidatePolicy(policy any) error {
	p, ok := asTokenBucketPolicy(policy)
	if !ok {
		return fmt.Errorf("token_bucket algorithm requires TokenBucketPolicy")
	}
	return p.Validate()
}

func (TokenBucketAlgorithm) BuildRedisArgs(policy any, requested int64) ([]any, error) {
	p, ok := asTokenBucketPolicy(policy)
	if !ok {
		return nil, fmt.Errorf("token_bucket algorithm requires TokenBucketPolicy")
	}
	if err := p.Validate(); err != nil {
		return nil, err
	}
	return []any{p.Capacity, p.RefillRate, requested, p.TTLSec}, nil
}

func (TokenBucketAlgorithm) ParseDecision(raw any) (Decision, error) {
	return parseNormalizedDecision(raw)
}

func asTokenBucketPolicy(policy any) (TokenBucketPolicy, bool) {
	switch p := policy.(type) {
	case TokenBucketPolicy:
		return p, true
	case *TokenBucketPolicy:
		if p == nil {
			return TokenBucketPolicy{}, false
		}
		return *p, true
	default:
		return TokenBucketPolicy{}, false
	}
}
